### PR TITLE
create `PnpInferenceState` with logprob and scores, implement (one-shot) sampling

### PR DESCRIPTION
**Pull request by joelgrus**
Create new object `PnpInferenceContext` that contains 

* the `CompGraph`
* a set of active `ExecutionScores`

and change everything to use it.

Also, the one-shot sampling should be working correctly. After this PR I'll take a stab at the particle-filter version.

### CODE REVIEW

**Comment by joelgrus** [comment]
ok, never mind this, I'm going to retool it some

**Comment by joelgrus** [comment]
ok, I think this version is more what you had in mind, let me know how you feel about it

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 1]
```diff
@@ -46,14 +46,18 @@ class BsoTrainer(val epochs: Int, val beamSize: Int, val maxIters: Int,
...
+    val queue = new BsoPnpQueue[A](beamSize, stateCost)
+    val finished = new BsoPnpQueue[A](beamSize, stateCost)
     val endContinuation = new PnpEndContinuation[A]()
 
-    val startEnv = env.setLog(log)
-    queue.offer(example.unconditional, env, 0.0, null, null, env)
+    val startEnv = env
+    // TODO(is this right?)
+    val context = PnpInferenceContext.init(model)
+        .addExecutionScore(stateCost)
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: Don't add the execution score here. (The way the scores are used here is slightly weird)

File: src/main/scala/org/allenai/pnp/BsoTrainer.scala [hunk 2]
```diff
@@ -179,28 +183,26 @@ class BsoTrainer(val epochs: Int, val beamSize: Int, val maxIters: Int,
...
-
+      context.log.stopTimer("bso/beam_search/search_step/eval_cost")
       val nextEnv = if (cost != 0.0) {
         env.setVar(EXECUTION_INCORRECT_VAR_NAME, null)
       } else {
         env
       }
-      
+
+      // TODO(joelgrus) is this right?
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/BsoTrainer.scala]: It is for the current implementation (which just uses the scores as a correct/incorrect signal). I have a fix in the ICCV PR that I'll send out soon.

File: src/main/scala/org/allenai/pnp/CompGraph.scala [hunk 3]
```diff
@@ -12,10 +12,10 @@ class CompGraph(val model: Model,
   
   // Initialize the nodes of the graph with a node per
   // parameter.
-  val paramExpressions = new Array[Expression](paramNames.size)
-  for (i <- 0 until paramNames.size) {
-    paramExpressions(i) = Expression.parameter(new Parameter(model, i))
-  }
+  //val paramExpressions = new Array[Expression](paramNames.size)
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/CompGraph.scala]: Go ahead and delete this dead code.

**Pull request merged**